cmd/bucket-policy-handlers.go

```diff
@@ -192,13 +192,38 @@ func (api objectAPIHandlers) PutBucketPolicyHandler(w http.ResponseWriter, r *ht
 		return
 	}
 
-	// Set the bucket policy in memory.
-	globalBucketPolicies.SetBucketPolicy(bucket, policy)
-
 	// Success.
 	writeSuccessNoContent(w)
 }
 
+// PutBucketPolicy - Persist the given policy to object layer, and
+// notify nodes in the cluster about the change. In-memory state is
+// updated in response to the notification. Assumes that the given
+// bucketPolicy is validated or is nil. If it is nil, bucket policy is
+// removed.
+func PutBucketPolicy(bucket string, bktpolicy *bucketPolicy, objAPI ObjectLayer) error {
+	// FIXME: Race exists between the bucket existence check and
+	// then updating the bucket policy.
+	if err := isBucketExist(bucket, objAPI); err != nil {
+		return err
+	}
+
+	// persist to object layer
+	if bktpolicy == nil {
```

Comment:
A PutBucketPolicy cannot have an empty body, this signifies that it can and we treat it like RemoveBucketPolicy.  This is not the same as bucket notification we shouldn't convey that in the flow. It is okay to have a separate remove bucket policy.  Any code here pretty much has to convey the S3 spec.